File: examples/go/transactions.go
package main

import (
	"bytes"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"log"
	"net/http"
	"net/url"
	"os"

	"github.com/joho/godotenv"
)

const transactionCollection = "transaction_accounts_go"

func getAuthToken() (string, error) {
	baseURL := os.Getenv("API_BASE_URL")
	apiKey := os.Getenv("API_BASE_KEY")
	authReq := map[string]string{"api_key": apiKey}
	body, _ := json.Marshal(authReq)

	resp, err := http.Post(baseURL+"/api/auth/token", "application/json", bytes.NewBuffer(body))
	if err != nil {
		return "", err
	}
	defer resp.Body.Close()

	responseBody, err := io.ReadAll(resp.Body)
	if err != nil {
		return "", err
	}
	if resp.StatusCode >= 300 {
		return "", fmt.Errorf("HTTP %d: %s", resp.StatusCode, responseBody)
	}
	if len(responseBody) == 0 {
		return "", fmt.Errorf("empty authentication response")
	}
	var result map[string]interface{}
	if err := json.Unmarshal(responseBody, &result); err != nil {
		return "", err
	}
	return result["token"].(string), nil
}

func makeRequest(method, path string, token string, data interface{}) (map[string]interface{}, error) {
	baseURL := os.Getenv("API_BASE_URL")
	var body io.Reader
	if data != nil {
		jsonData, _ := json.Marshal(data)
		body = bytes.NewBuffer(jsonData)
	}

	req, err := http.NewRequest(method, baseURL+path, body)
	if err != nil {
		return nil, err
	}

	req.Header.Set("Authorization", "Bearer "+token)
	req.Header.Set("Content-Type", "application/json")

	client := &http.Client{}
	resp, err := client.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	responseBody, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}
	if resp.StatusCode >= 300 && resp.StatusCode != http.StatusNotFound {
		return nil, fmt.Errorf("HTTP %d: %s", resp.StatusCode, responseBody)
	}
	if resp.StatusCode == http.StatusNotFound {
		return nil, nil
	}
	if len(responseBody) == 0 {
		return map[string]interface{}{}, nil
	}
	var result map[string]interface{}
	if err := json.Unmarshal(responseBody, &result); err != nil {
		return nil, err
	}
	return result, nil
}

func run() (runErr error) {
	// Load environment variables
	err := godotenv.Load()
	if err != nil {
		return fmt.Errorf("load environment variables: %w", err)
	}

	// Get authentication token
	token, err := getAuthToken()
	if err != nil {
		return err
	}
	fmt.Print("✓ Authentication successful\n\n")
	activeTransactions := make(map[string]bool)
	defer func() {
		for txID := range activeTransactions {
			if _, err := makeRequest("POST", "/api/transactions/"+txID+"/rollback", token, nil); err != nil {
				runErr = errors.Join(runErr, fmt.Errorf("rollback transaction %s during cleanup: %w", txID, err))
			}
		}
		if _, err := makeRequest("DELETE", "/api/collections/"+transactionCollection, token, nil); err != nil {
			runErr = errors.Join(runErr, fmt.Errorf("cleanup collection %s: %w", transactionCollection, err))
		}
	}()

	// Setup test data
	fmt.Println("=== Setup: Creating Test Accounts ===")
	alice, err := makeRequest("POST", "/api/insert/"+transactionCollection, token, map[string]interface{}{
		"account_id": "ACC001",
		"name":       "Alice",
		"balance":    1000,
	})
	if err != nil {
		return err
	}
	aliceID := alice["id"].(string)
	fmt.Printf("Created Alice: $1000 - ID: %s\n", aliceID)

	bob, err := makeRequest("POST", "/api/insert/"+transactionCollection, token, map[string]interface{}{
		"account_id": "ACC002",
		"name":       "Bob",
		"balance":    500,
	})
	if err != nil {
		return err
	}
	bobID := bob["id"].(string)
	fmt.Printf("Created Bob: $500 - ID: %s\n\n", bobID)

	// Example 1: Begin Transaction
	fmt.Println("=== Example 1: Begin Transaction ===")
	txData, err := makeRequest("POST", "/api/transactions", token, map[string]interface{}{
		"isolation_level": "ReadCommitted",
	})
	if err != nil {
		return err
	}
	txID := txData["transaction_id"].(string)
	activeTransactions[txID] = true
	fmt.Printf("Transaction ID: %s\n\n", txID)

	// Example 2: Operations with transaction_id
	fmt.Println("=== Example 2: Operations with transaction_id ===")
	_, err = makeRequest("PUT", "/api/batch/update/"+transactionCollection+"?transaction_id="+url.QueryEscape(txID), token, map[string]interface{}{
		"updates": []map[string]interface{}{
			{"id": aliceID, "data": map[string]interface{}{"balance": 800}},
			{"id": bobID, "data": map[string]interface{}{"balance": 700}},
		},
	})
	if err != nil {
		return err
	}
	fmt.Println("Updated Alice: $1000 → $800")
	fmt.Print("Updated Bob: $500 → $700\n\n")

	// Example 3: Transaction Status
	fmt.Println("=== Example 3: Transaction Status ===")
	status, err := makeRequest("GET", "/api/transactions/"+txID, token, nil)
	if err != nil {
		return err
	}
	fmt.Printf("Status: %v\n", status["state"])
	fmt.Printf("Operations: %.0f\n\n", status["operations_count"])

	// Example 4: Commit Transaction
	fmt.Println("=== Example 4: Commit Transaction ===")
	_, err = makeRequest("POST", "/api/transactions/"+txID+"/commit", token, nil)
	if err != nil {
		return err
	}
	delete(activeTransactions, txID)
	fmt.Print("✓ Transaction committed\n\n")

	// Verification
	fmt.Println("=== Verification ===")
	aliceFinal, err := makeRequest("GET", "/api/find/"+transactionCollection+"/"+aliceID, token, nil)
	if err != nil {
		return err
	}
	fmt.Printf("Alice: %v\n", aliceFinal["balance"])

	bobFinal, err := makeRequest("GET", "/api/find/"+transactionCollection+"/"+bobID, token, nil)
	if err != nil {
		return err
	}
	fmt.Printf("Bob: %v\n\n", bobFinal["balance"])

	// Example 5: Rollback
	fmt.Println("=== Example 5: Rollback ===")
	txData2, err := makeRequest("POST", "/api/transactions", token, map[string]interface{}{
		"isolation_level": "ReadCommitted",
	})
	if err != nil {
		return err
	}
	txID2 := txData2["transaction_id"].(string)
	activeTransactions[txID2] = true
	fmt.Printf("New transaction: %s\n", txID2)

	_, err = makeRequest("PUT", "/api/batch/update/"+transactionCollection+"?transaction_id="+url.QueryEscape(txID2), token, map[string]interface{}{
		"updates": []map[string]interface{}{
			{"id": bobID, "data": map[string]interface{}{"balance": 600}},
		},
	})
	if err != nil {
		return err
	}
	fmt.Println("Updated Bob: $700 → $600 (in transaction)")

	_, err = makeRequest("POST", "/api/transactions/"+txID2+"/rollback", token, nil)
	if err != nil {
		return err
	}
	delete(activeTransactions, txID2)
	fmt.Println("✓ Transaction rolled back")

	bobAfter, err := makeRequest("GET", "/api/find/"+transactionCollection+"/"+bobID, token, nil)
	if err != nil {
		return err
	}
	fmt.Printf("Bob after rollback: %v\n\n", bobAfter["balance"])

	// Cleanup
	fmt.Println("=== Cleanup ===")
	if _, err := makeRequest("DELETE", "/api/delete/"+transactionCollection+"/"+aliceID, token, nil); err != nil {
		return fmt.Errorf("delete Alice: %w", err)
	}
	if _, err := makeRequest("DELETE", "/api/delete/"+transactionCollection+"/"+bobID, token, nil); err != nil {
		return fmt.Errorf("delete Bob: %w", err)
	}
	fmt.Print("✓ Deleted test accounts\n\n")

	fmt.Println("✓ All transaction examples completed")
	return nil
}

func main() {
	if err := run(); err != nil {
		log.Fatal(err)
	}
}

File: examples/kotlin/transactions.kt
import io.github.cdimascio.dotenv.dotenv
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONArray
import org.json.JSONObject

val client = OkHttpClient()
val JSON = "application/json; charset=utf-8".toMediaType()

fun getAuthToken(): String {
    val dotenv = dotenv()
    val baseURL = dotenv["API_BASE_URL"] ?: "http://localhost:8080"
    val apiKey = dotenv["API_BASE_KEY"] ?: "a-test-api-key-from-ekodb"
    
    val json = JSONObject().put("api_key", apiKey).toString()
    val body = json.toRequestBody(JSON)
    val request = Request.Builder()
        .url("$baseURL/api/auth/token")
        .post(body)
        .build()

    client.newCall(request).execute().use { response ->
        val jsonResponse = JSONObject(response.body!!.string())
        return jsonResponse.getString("token")
    }
}

fun makeRequest(method: String, path: String, token: String, data: JSONObject? = null): JSONObject {
    val dotenv = dotenv()
    val baseURL = dotenv["API_BASE_URL"] ?: "http://localhost:8080"
    
    val requestBuilder = Request.Builder()
        .url("$baseURL$path")
        .header("Authorization", "Bearer $token")
        .header("Content-Type", "application/json")

    when (method) {
        "POST" -> requestBuilder.post(data?.toString()?.toRequestBody(JSON) ?: "{}".toRequestBody(JSON))
        "PUT" -> requestBuilder.put(data?.toString()?.toRequestBody(JSON) ?: "{}".toRequestBody(JSON))
        "DELETE" -> requestBuilder.delete()
        else -> requestBuilder.get()
    }

    val request = requestBuilder.build()
    client.newCall(request).execute().use { response ->
        val responseBody = response.body!!.string()
        return if (responseBody.isNotEmpty()) JSONObject(responseBody) else JSONObject()
    }
}

fun main() {
    val token = getAuthToken()
    println("✓ Authentication successful\n")

    // Setup test data
    println("=== Setup: Creating Test Accounts ===")
    val alice = makeRequest("POST", "/api/insert/test_accounts", token, JSONObject().apply {
        put("account_id", "ACC001")
        put("name", "Alice")
        put("balance", 1000)
    })
    val aliceId = alice.getString("id")
    println("Created Alice: \$1000 - ID: $aliceId")

    val bob = makeRequest("POST", "/api/insert/test_accounts", token, JSONObject().apply {
        put("account_id", "ACC002")
        put("name", "Bob")
        put("balance", 500)
    })
    val bobId = bob.getString("id")
    println("Created Bob: \$500 - ID: $bobId\n")

    // Example 1: Begin Transaction
    println("=== Example 1: Begin Transaction ===")
    val txData = makeRequest("POST", "/api/transactions", token, JSONObject().apply {
        put("isolation_level", "ReadCommitted")
    })
    val txId = txData.getString("transaction_id")
    println("Transaction ID: $txId\n")

    // Example 2: Operations with transaction_id
    println("=== Example 2: Operations with transaction_id ===")
    makeRequest("PUT", "/api/batch/update/test_accounts?transaction_id=$txId", token, JSONObject().apply {
        put("updates", JSONArray().apply {
            put(JSONObject().apply {
                put("id", aliceId)
                put("data", JSONObject().put("balance", 800))
            })
            put(JSONObject().apply {
                put("id", bobId)
                put("data", JSONObject().put("balance", 700))
            })
        })
    })
    println("Updated Alice: \$1000 → \$800")
    println("Updated Bob: \$500 → \$700\n")

    // Example 3: Transaction Status
    println("=== Example 3: Transaction Status ===")
    val status = makeRequest("GET", "/api/transactions/$txId", token)
    println("Status: ${status.getString("state")}")
    println("Operations: ${status.getInt("operations_count")}\n")

    // Example 4: Commit Transaction
    println("=== Example 4: Commit Transaction ===")
    makeRequest("POST", "/api/transactions/$txId/commit", token)
    println("✓ Transaction committed\n")

    // Verification
    println("=== Verification ===")
    val aliceFinal = makeRequest("GET", "/api/find/test_accounts/$aliceId", token)
    println("Alice: ${aliceFinal.get("balance")}")

    val bobFinal = makeRequest("GET", "/api/find/test_accounts/$bobId", token)
    println("Bob: ${bobFinal.get("balance")}\n")

    // Example 5: Rollback
    println("=== Example 5: Rollback ===")
    val txData2 = makeRequest("POST", "/api/transactions", token, JSONObject().apply {
        put("isolation_level", "ReadCommitted")
    })
    val txId2 = txData2.getString("transaction_id")
    println("New transaction: $txId2")

    makeRequest("PUT", "/api/batch/update/test_accounts?transaction_id=$txId2", token, JSONObject().apply {
        put("updates", JSONArray().apply {
            put(JSONObject().apply {
                put("id", bobId)
                put("data", JSONObject().put("balance", 600))
            })
        })
    })
    println("Updated Bob: \$700 → \$600 (in transaction)")

    makeRequest("POST", "/api/transactions/$txId2/rollback", token)
    println("✓ Transaction rolled back")

    val bobAfter = makeRequest("GET", "/api/find/test_accounts/$bobId", token)
    println("Bob after rollback: ${bobAfter.get("balance")}\n")

    // Cleanup
    println("=== Cleanup ===")
    makeRequest("DELETE", "/api/delete/test_accounts/$aliceId", token)
    makeRequest("DELETE", "/api/delete/test_accounts/$bobId", token)
    println("✓ Deleted test accounts\n")

    println("✓ All transaction examples completed")
}
